enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+Machines. For example, we expect on AWS that this will contain a list of availability zones within a single region.
+Note, we are explicitly not expecting users to add different regions to the the `FailureDomains` as we do not support
+running OpenShift across multiple regions.
+
+The following table denotes the field on each platform into which the list of failure domains will be mapped:
+
+
+| Platform  | ProviderSpec Field          |
+| --------- | --------------------------- |
+| AWS       | .placement.availabilityZone |
+| Azure     | .zone                       |
+| GCP       | .zone                       |
+| vSphere   | TBD (see below)             |
+| OpenStack | .availabilityZone           |
+
+Note that on some platforms (eg OpenStack) the failure domain field is optional, as such, the `FailureDomains` field
+must also be optional.
+
+###### Failure Domains on vSphere
+
+As of OpenShift 4.10, there is no concept of a zone within vSphere and as such we would expect that the `FailureDomains`
+would be omitted for vSphere.
+
+However, [there is future work](https://github.com/openshift/enhancements/pull/918) to include zone support for vSphere
+within 4.11. Once this is implemented, the Machine API provider for vSphere will understand the concept of zones,
+defined within the `Infrastructure` resource.
+The `ControlPlaneMachineSet` will then be able to balance Machines across the different zones by setting the new zone
+information within the provider specs as it does on other platforms.
+
+#### Ensuring Machines match the desired state
+
+To ensure Machines are up to date with the desired configuration, the `ControlPlaneMachineSet` controller will leverage
+a similar pattern to the workload controllers.
+It will hash the template and compare the hashed result with the spec of the Machines within the managed set.
+As we expect the failure domain to vary across the managed Machines, this will need to be omitted before the hash can
+be calculated.
+
+Should any Machine not match the desired hash, it will be updated based on the chosen update strategy.
+
+##### The RollingUpdate update strategy
+
+The RollingUpdate strategy will be the default. It mirrors the RollingUpdate strategy familiar to users from
+Deployments. When a change is required to be rolled out, the Machine controller will create a new Control Plane
+Machine, wait for this to become ready, and then remove the old Machine. It will repeat this process until all Machines
+are up to date.
+During this process the etcd membership will be protected by the mechanism described in a [separate enhancement](https://github.com/openshift/enhancements/pull/943),
+in particular it isn't essential that the Control Plane Machines are updated in a rolling fashion for etcd sakes, though
+to avoid potential issues with the spread of etcd members across availability zones during update, the
+`ControlPlaneMachineSet` will perform a rolling update zone by zone.
```

Comment:
```suggestion
to avoid potential issues with the spread of etcd members across failure domains during update, the
`ControlPlaneMachineSet` will perform a rolling update domain by domain.
```